### PR TITLE
Link to the SonarQube dashboard on Maven if SQ integration is enabled.

### PR DESCRIPTION
**Pull request by HSAR**

### CODE REVIEW

File: Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts [hunk 1]
```diff
@@ -67,28 +65,42 @@ export function uploadCodeAnalysisResults():void {
 }
 
 // Returns the names of any enabled code analysis tools, or empty array if none.
-function getEnabledCodeAnalysisTools():string[] {
-    var result:string[] = [];
+function getEnabledCodeAnalysisTools(): Set<string> {
```
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts]: ECMA6 has a language-level Set implementation, which is far more appropriate in this case.

File: Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts [hunk 2]
```diff
@@ -67,28 +65,42 @@ export function uploadCodeAnalysisResults():void {
...
 
     return result;
 }
 
 // Returns true if the given code analysis tool is enabled
-function isCodeAnalysisToolEnabled(toolName:string) {
+export function isCodeAnalysisToolEnabled(toolName:string): boolean {
     // Get the list of enabled tools, return whether or not toolName is contained in it
-    return (getEnabledCodeAnalysisTools().indexOf(toolName) > -1);
+    var result = getEnabledCodeAnalysisTools().has(toolName);
```
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts]: Changed some of the logging to reduce clutter. Previously we used to print out "XYZ is enabled" on every call to getEnabledAnalysisTools(), which was not great. Now we only produce it in the specific call, and it is localised.

File: Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts [hunk 3]
```diff
@@ -67,28 +65,42 @@ export function uploadCodeAnalysisResults():void {
...
+    return result;
 }
 
 // Returns the full path of the staging directory for a given tool.
-function getStagingDirectory(toolName:string):string {
+function getToolStagingDirectory(toolName:string):string {
     return path.join(stagingDir, toolName.toLowerCase());
 }
 
+export function getMasterStagingDirectory() {
```
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts]: mavensonar requires access to this directory.

File: Tasks/Maven/CodeAnalysis/mavenpmd.ts [hunk 4]
```diff
@@ -28,6 +28,10 @@ export function collectPmdOutput(rootDir:string) : AnalysisResult {
 
     var pmdXmlFilePath = path.join(rootDir, 'target', 'pmd.xml');
     result = collectPmdXml(result, pmdXmlFilePath);
+    if (!result)
```
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavenpmd.ts]: Additional defensive coding was required here and in some other places, since the SQ plugin creates target folders that do not have pmd.xml files.

File: Tasks/Maven/CodeAnalysis/mavensonar.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
+/// <reference path="../../../definitions/vsts-task-lib.d.ts" />
```
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavensonar.ts]: This file consists of  two methods moved here from maventask, and one new method to generate and upload the summary.

The moved methods have had additional arguments added, since they cannot access some variables previously global inside maventask.

File: Tasks/Maven/maventask.ts [hunk 6]
```diff
@@ -167,16 +168,22 @@ mvnGetVersion.exec()
...
-            // Run Maven with the sonar:sonar goal, even if the user-goal Maven failed (e.g. test failures).
-            // Note that running sonar:sonar along with the user goals is not supported due to a SonarQube bug.
-            return mvnsq.exec()
+        if (codeAnalysis.isCodeAnalysisToolEnabled(sqMaven.toolName)) {
+            var mvnsq:ToolRunner = sqMaven.getSonarQubeRunner(mvnExec, mavenPOMFile, mavenOptions, execFileJacoco);
+            var result;
+            if (mvnsq) {
+                // Run Maven with the sonar:sonar goal, even if the user-goal Maven failed (e.g. test failures).
+                // Note that running sonar:sonar along with the user goals is not supported due to a SonarQube bug.
+                mvnsq.exec().then((code) => {
```
Comment (HSAR) [inline Tasks/Maven/maventask.ts]: Promise used to make sure we don't have any timing issues.

File: Tasks/Common/sonarqube-common/sonarqube-common.ts [hunk 7]
```diff
@@ -87,6 +90,28 @@ export function getSonarQubeEndpointFromInput(inputFieldName): SonarQubeEndpoint
     return new SonarQubeEndpoint(hostUrl, hostUsername, hostPassword);
 }
 
+// Returns, as an object, the contents of the 'report-task.txt' file created by SonarQube plugins
+// The returned object contains the following properties:
+//   projectKey, serverUrl, dashboardUrl, ceTaskId, ceTaskUrl
+export function getSonarQubeTaskReport(sonarPluginFolder: string) {
+    var reportFilePath:string = path.join(sonarPluginFolder, 'report-task.txt');
```
Comment (bgavrilMS) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: This function does too much - it computes the location of the file and then reads it. I recommend you let the responsability of computing the path to a different function

File: Tasks/Common/sonarqube-common/sonarqube-common.ts [hunk 8]
```diff
@@ -87,6 +90,28 @@ export function getSonarQubeEndpointFromInput(inputFieldName): SonarQubeEndpoint
...
+    if (!tl.exist(reportFilePath)) {
+        tl.debug('Task report not found at: ' + reportFilePath);
```
Comment (bgavrilMS) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: Do we expect the path to not be there? I think this should be at least a warning or an exception.

File: Tasks/Common/sonarqube-common/sonarqube-common.ts [hunk 9]
```diff
@@ -87,6 +90,28 @@ export function getSonarQubeEndpointFromInput(inputFieldName): SonarQubeEndpoint
...
+        return null;
+    }
+
+    var reportFileString: string = fs.readFileSync(reportFilePath, 'utf-8');
+    var reportLines: string[] = reportFileString.replace(/\r\n/g, '\n').split('\n'); // proofs against xplat line-ending issues
+
+    var reportObject = {};
```
Comment (bgavrilMS) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: Consider using a strongly typed construct, like an ES6 map or  [key: string]: string;

File: Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts [hunk 10]
```diff
@@ -67,28 +65,42 @@ export function uploadCodeAnalysisResults():void {
...
 
 // Returns true if the given code analysis tool is enabled
-function isCodeAnalysisToolEnabled(toolName:string) {
+export function isCodeAnalysisToolEnabled(toolName:string): boolean {
     // Get the list of enabled tools, return whether or not toolName is contained in it
-    return (getEnabledCodeAnalysisTools().indexOf(toolName) > -1);
+    var result = getEnabledCodeAnalysisTools().has(toolName);
+    if (result) {
+        // Looks like: 'SonarQube analysis is enabled.'
+        console.log(tl.loc('codeAnalysis_toolIsEnabled', toolName));
```
Comment (bgavrilMS) [inline Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts]: This function seems to be general purpose and I'd expect it to be called many times. However it has a side effect - it also logs that "tool X is enabled". I recommend you move the logging elsewhere, higher up.

File: Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts (hunk 3 above)
Comment (bgavrilMS) [inline Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts]: Please rename as the word "master" has too many meanings. Perhaps "code analysis staging dir" and "tool staging dir" ?

File: Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts [hunk 11]
```diff
@@ -29,26 +31,22 @@ export function applyEnabledCodeAnalysisGoals(mvnRun: trm.ToolRunner):trm.ToolRu
 // Extract data from code analysis output files and upload results to build server
 export function uploadCodeAnalysisResults():void {
     // Return early if no analysis tools are enabled
-    var enabledCodeAnalysisTools = getEnabledCodeAnalysisTools();
-    if (enabledCodeAnalysisTools.length < 1) {
+    var enabledCodeAnalysisTools: Set<string> = getEnabledCodeAnalysisTools();
```
Comment (bgavrilMS) [inline Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts]: In the future, consider using classes and public / private functions to organise your modules. OO design is easier to maintain in the long run.

File: Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts (hunk 1 above)
Comment (bgavrilMS) [inline Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts]: Please group all public (i.e. export) functions to the top of the file and all private functions to the bottom

File: Tasks/Maven/CodeAnalysis/mavenpmd.ts (hunk 4 above)
Comment (bgavrilMS) [inline Tasks/Maven/CodeAnalysis/mavenpmd.ts]: Code might be easier to read if you invert this - if (result) { do stuff }

File: Tasks/Maven/CodeAnalysis/mavensonar.ts [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+// Upload a build summary with links to available SonarQube dashboards for further analysis details.
+export function uploadSonarQubeBuildSummary(): void {
+    // Collect all unique dashboard URLs
+    // project key -> dashboard URL
+    var dashboardUrlsByProjectKey: Map<string, string> = new Map<string, string>();
+
+    var sourcesDir = tl.getVariable('build.sourcesDirectory');
+    var modules:ModuleAnalysis[] = codeAnalysis.findCandidateModules(sourcesDir);
+    modules.forEach((module: ModuleAnalysis) => {
```
Comment (bgavrilMS) [inline Tasks/Maven/CodeAnalysis/mavensonar.ts]: My understanding is that SQ analysis is orthogonal to multi-module Maven / Gradle. Why are there multiple dashboards?

File: Tasks/Maven/CodeAnalysis/mavensonar.ts [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        var taskReport = sqCommon.getSonarQubeTaskReport(path.join(module.rootDirectory, 'target', 'sonar'));
+        if (taskReport) {
+            dashboardUrlsByProjectKey.set(taskReport['projectKey'], taskReport['dashboardUrl']);
+        }
+    });
+
+    dashboardUrlsByProjectKey.delete(null);
```
Comment (bgavrilMS) [inline Tasks/Maven/CodeAnalysis/mavensonar.ts]: do not add nulls in the first place.

File: Tasks/Maven/CodeAnalysis/mavensonar.ts [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+           lineBodyText = tl.loc('sqAnalysisBuildSummaryLine_OneProject');
+       }
+
+       buildSummaryLines.push(util.format('[%s >](%s "%s Dashboard")',
+           lineBodyText, dashboardUrlsByProjectKey.get(projectKey), projectKey));
+   };
+
+    // Save and upload build summary
+    // Looks like: "[Detailed SonarQube report (foo) >](https://mySQserver:9000/dashboard/index/foo "foo Dashboard")  \r\n
+    // [Detailed SonarQube report (bar) >](https://mySQserver:9000/dashboard/index/bar "bar Dashboard")  \r\n"
```
Comment (bgavrilMS) [inline Tasks/Maven/CodeAnalysis/mavensonar.ts]: Split this into a separate func

**Comment by bgavrilMS** [comment]
There are no tests. We don't  usually ask for code reviews unless test are written, because the code could significantly change to allow for test hooks, bug fixes etc.

File: Tasks/Common/sonarqube-common/sonarqube-common.ts [hunk 15]
```diff
@@ -1,11 +1,14 @@
...
+import path = require('path');
+import fs = require('fs');
+
 import tl = require('vsts-task-lib/task');
 import {ToolRunner} from 'vsts-task-lib/toolrunner';
 
 // Simple data class for a SonarQube generic endpoint
 export class SonarQubeEndpoint {
-    constructor(public Url, public Username, public Password) {
+    constructor(public Url: string, public Username: string, public Password: string) {
```
Comment (HSAR) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: Changed for neatness: added types.

File: Tasks/Common/sonarqube-common/sonarqube-common.ts (hunk 8 above)
Comment (HSAR) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: The precursor to this function, findCandidateModules, has a high rate of false positives (especially if there is more than one code analysis tool running, i.e. PMD), so this is not necessarily a problem.

I will, however, add a warning for if no task reports are found at all.

Accepted, in progress.

File: Tasks/Common/sonarqube-common/sonarqube-common.ts (hunk 7 above)
Comment (HSAR) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: Accepted, to do.

Comment (HSAR) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: Resolved, done.

It seemed better to hide the implementation of how the parsing is done, so I've moved that into a seperate method instead of the computing path stuff.

File: Tasks/Common/sonarqube-common/sonarqube-common.ts (hunk 8 above)
Comment (HSAR) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: Resolved, added a localized warning message.

File: Tasks/Common/sonarqube-common/sonarqube-common.ts (hunk 9 above)
Comment (HSAR) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: Resolved, done.

File: Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts (hunk 11 above)
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts]: Resolved, done.

File: Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts (hunk 1 above)
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts]: Resolved, done.

File: Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts (hunk 10 above)
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts]: I think this is a good place to be logging it. Then we get a log message every time one of these tools finds it is enabled, and no message if they do not. Both are useful pieces of debug information, and moving the logging upwards would result in a lot of identical method calls.

File: Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts (hunk 3 above)
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts]: Resolved, done.

File: Tasks/Maven/CodeAnalysis/mavenpmd.ts (hunk 4 above)
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavenpmd.ts]: Discussed offline. This way the code makes it clear that a null return from _collectPmdXML()_ results in a null return of its own.

File: Tasks/Maven/CodeAnalysis/mavensonar.ts (hunk 12 above)
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavensonar.ts]: Discussed offline - the main reason for doing this is not specifically for multi-module, but to support master POM files that are not in the root (i.e. expected) directory. SonarQube creates its report file in _root/target/sonar/_ directory.

However, since this seems to add substantial complexity for an edge case, I'll remove it in favour of a simplified approach.

File: Tasks/Maven/CodeAnalysis/mavensonar.ts (hunk 13 above)
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavensonar.ts]: This was a defense against the now-removed implementation where use of an object could easily result in stray nulls being added. This can (and will) now be safely removed.

Accepted, in progress.

File: Tasks/Maven/CodeAnalysis/mavensonar.ts (hunk 12 above)
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavensonar.ts]: Resolved, done.

File: Tasks/Maven/CodeAnalysis/mavensonar.ts (hunk 13 above)
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavensonar.ts]: Resolved, done.

File: Tasks/Maven/CodeAnalysis/mavensonar.ts (hunk 14 above)
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavensonar.ts]: Function now has the following sections:

- Parse
- Create summary
- Save to disk
- Upload

Resolved, done.

File: Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts (hunk 1 above)
Comment (HSAR) [inline Tasks/Maven/CodeAnalysis/mavencodeanalysis.ts]: Resolved, done.

File: Tasks/Common/sonarqube-common/sonarqube-common.ts [hunk 16]
```diff
@@ -87,6 +90,19 @@ export function getSonarQubeEndpointFromInput(inputFieldName): SonarQubeEndpoint
     return new SonarQubeEndpoint(hostUrl, hostUsername, hostPassword);
 }
 
+// Returns, as an object, the contents of the 'report-task.txt' file created by SonarQube plugins
+// The returned object contains the following properties:
+//   projectKey, serverUrl, dashboardUrl, ceTaskId, ceTaskUrl
+export function getSonarQubeTaskReport(sonarPluginFolder: string): Map<string, string> {
```
Comment (bgavrilMS) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: In hindsight, why use a map at all when you can have an actual object?

File: Tasks/Maven/CodeAnalysis/mavensonar.ts [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+                                    sqDbUrl?: string, sqDbUsername?: string, sqDbPassword?: string) {
+    var mvnsq = tl.createToolRunner(mvnPath);
+
+    mvnsq = sqCommon.applySonarQubeConnectionParams(mvnsq, sqHostUrl, sqHostUsername, sqHostPassword, sqDbUrl, sqDbUsername, sqDbPassword);
+
+    return mvnsq;
+}
+
+// Creates the string that comprises the build summary text.
+function createSonarQubeBuildSummary(taskReport: Map<string, string>): string {
```
Comment (bgavrilMS) [inline Tasks/Maven/CodeAnalysis/mavensonar.ts]: There's a chance that the format of that file will change. It's better to fail in that case, i.e. be more defensive (this should be done when you parse the file though, not here).

File: Tests/L0/Maven/_suite.ts [hunk 18]
```diff
@@ -684,23 +685,33 @@ describe('Maven Suite', function() {
             });
     })
 
-    it('run maven including SonarQube analysis', (done) => {
-       
-       // not a valid PR branch
+    it('run maven including SonarQube analysis', function(done) {
```
Comment (bgavrilMS) [inline Tests/L0/Maven/_suite.ts]: As discussed, you should aim for your test to exercise as much as possible of the written code. At a minimum there should be 2 tests - one that starts from a valid task-report.txt and you then assert that your summary.md file contains the dashboard URL. The other one should cover having an invalid task-report.txt.

File: Tasks/Common/sonarqube-common/sonarqube-common.ts (hunk 16 above)
Comment (HSAR) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: We actually use both the dashboard URL and the project title (for the hover text). Using a map is more flexible and is designed specifically to be resistant against file changes. A data class might fulfil many of the same functions (though an implementation of that would probably still go through the Map just for neatness).

Comment (HSAR) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: Thinking about it a bit more, I think a data class would be much better. Then we can keep the exact names of the key/value pairs that correspond to its fields in the data class rather than in the method that creates the build summary.

Accept, in progress.

Comment (HSAR) [inline Tasks/Common/sonarqube-common/sonarqube-common.ts]: Resolved, done.

**Comment by HSAR** [comment]
Will reopen this PR - GitHub seems to have screwed up the diff.
